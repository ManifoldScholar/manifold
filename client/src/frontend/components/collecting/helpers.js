import flatMapDepth from "lodash/flatMapDepth";
import identity from "lodash/identity";
import has from "lodash/has";
import get from "lodash/get";
import isEmpty from "lodash/isEmpty";

export function getEntityCollection(entity, relationship = "collection") {
  return get(entity, `relationships.${relationship}`);
}

function isCollection(possibleCollection) {
  return has(possibleCollection, "attributes.categoryMappings");
}

function hasCollection(possibleEntity, relationship = "collection") {
  return has(possibleEntity, `relationships.${relationship}`);
}

export function collectedIdsForCollection(collection) {
  if (!collection) return [];
  const mappings = collection.attributes.categoryMappings;
  return flatMapDepth(
    Object.values(mappings).map(mapping => Object.values(mapping)),
    identity,
    2
  );
}

export function inCollection(collectionOrEntity, collectable) {
  const collection = isCollection(collectionOrEntity)
    ? collectionOrEntity
    : hasCollection(collectionOrEntity) &&
      getEntityCollection(collectionOrEntity);
  if (!collection) return false;

  const collectedIds = collectedIdsForCollection(collection);
  return collectedIds.includes(collectable.id);
}

export function inCollections(collectable, ...collectionsOrEntities) {
  return collectionsOrEntities.some(collectionOrEntity =>
    inCollection(collectionOrEntity, collectable)
  );
}

export function getResponse(id, responses) {
  return responses.find(res => res.id === id);
}

export function idInResponses(id, responses) {
  if (responses?.length < 1) return false;
  return responses.map(res => res.id).includes(id);
}

export function hasItemsInCollection(entity) {
  const collection = getEntityCollection(entity);
  if (!collection?.attributes) return false;
  return !isEmpty(collection.attributes.categoryMappings);
}
