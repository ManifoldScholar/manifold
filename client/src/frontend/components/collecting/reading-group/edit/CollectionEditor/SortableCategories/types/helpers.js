import PropTypes from "prop-types";

export const collectedShape = {
  categoryId: PropTypes.string.isRequired,
  collectedIds: PropTypes.array.isRequired,
  responses: PropTypes.array.isRequired,
  onRemove: PropTypes.func.isRequired,
  showDropzone: PropTypes.bool
};
