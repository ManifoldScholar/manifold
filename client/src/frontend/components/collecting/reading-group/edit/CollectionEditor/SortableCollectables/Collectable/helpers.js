export function getCollectableIcon(type) {
  switch (type) {
    case "projects":
      return "projects64";
    case "texts":
      return "textsLoosePages64";
    case "textSections":
      return "toc64";
    case "resourceCollections":
      return "resourceCollection64";
    case "resources":
      return "resources64";
    case "journalIssues":
      return "journals64";
    default:
      return "resources64";
  }
}
