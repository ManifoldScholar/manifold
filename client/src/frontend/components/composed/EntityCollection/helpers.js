import isEmpty from "lodash/isEmpty";

export function getHeroImage(collection) {
  const { heroLayout, heroStyles } = collection.attributes;
  if (isEmpty(heroStyles)) return null;
  switch (heroLayout) {
    case "full_bleed":
    case "wide_inset":
      return heroStyles.largeLandscape;
    case "square_inset":
      return heroStyles.mediumSquare;
    default:
      return null;
  }
}

export function getHeaderLayout(collection) {
  const {
    heroLayout,
    heroStyles,
    descriptionFormatted: description
  } = collection.attributes;
  if (isEmpty(heroStyles))
    return description ? "title_description" : "title_only";
  return heroLayout;
}
