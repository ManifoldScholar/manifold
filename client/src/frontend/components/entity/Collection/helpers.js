export function getHeroImage(layout, collection) {
  const { heroStyles } = collection?.attributes ?? {};
  switch (layout) {
    case "full_bleed":
    case "wide_inset":
      return heroStyles.largeLandscape;
    case "square_inset":
      return heroStyles.mediumSquare;
    default:
      return null;
  }
}

export function getHeaderLayout(collection) {
  const { heroLayout, heroStyles, descriptionFormatted: description } =
    collection?.attributes ?? {};
  if (heroLayout && heroStyles) return heroLayout;
  if (heroStyles?.largeLandscape) return "square_inset";
  return description ? "title_description" : "title_only";
}
