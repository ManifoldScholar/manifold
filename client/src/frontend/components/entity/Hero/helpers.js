import orderBy from "lodash/orderBy";

export const getAuth = (entity, authorization) => {
  const showErrors = authorization.current.authorizeAbility({
    entity,
    ability: "update"
  });
  const authorized = authorization.current.authorizeAbility({
    entity,
    ability: "fullyRead"
  });
  return { showErrors, authorized };
};

export const getPartsData = entity => {
  const callouts = entity.relationships?.actionCallouts ?? [];
  const orderedCallouts = orderBy(
    callouts,
    ["attributes.button", "attributes.location", "attributes.position"],
    ["desc", "asc", "asc"]
  );
  const calloutsBySide = side =>
    orderBy(
      callouts.filter(callout => callout.attributes.location === side),
      ["attributes.button", "attributes.position"],
      ["desc", "asc"]
    );
  const leftCallouts = calloutsBySide("left");
  const rightCallouts = calloutsBySide("right");

  const copy = entity.attributes.imageCreditsFormatted;
  const bgImage =
    entity.attributes.heroStyles?.largeLandscape &&
    entity.attributes.heroStyles?.mediumLandscape
      ? entity.attributes.heroStyles
      : false;
  const bgAlt = entity.attributes.heroAltText;

  const twitter = entity.attributes.twitterId;
  const instagram = entity.attributes.instagramId;
  const facebook = entity.attributes.facebookId;
  const hashtag = entity.attributes.hashtag;
  const social = !!twitter || !!instagram || !!facebook || !!hashtag;

  const description = entity.attributes.descriptionFormatted;
  const { creators, contributors } = entity.relationships;
  const cover = entity.attributes.coverStyles?.medium;
  return {
    callouts,
    orderedCallouts,
    leftCallouts,
    rightCallouts,
    copy,
    bgImage,
    bgAlt,
    twitter,
    instagram,
    facebook,
    hashtag,
    social,
    description,
    creators,
    contributors,
    cover
  };
};
