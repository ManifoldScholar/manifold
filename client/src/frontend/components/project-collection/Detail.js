import React, { Component } from "react";
import PropTypes from "prop-types";
import ProjectList from "frontend/components/project-list";
import Utility from "global/components/utility";
import get from "lodash/get";
import Header from "./Header";
import Filters from "./Filters";

export default class ProjectCollectionDetail extends Component {
  static displayName = "ProjectCollectionDetail";

  static propTypes = {
    projectCollection: PropTypes.object,
    authentication: PropTypes.object,
    dispatch: PropTypes.func,
    pagination: PropTypes.object,
    paginationClickHandler: PropTypes.func,
    projects: PropTypes.array,
    filterChangeHandler: PropTypes.func,
    initialState: PropTypes.object
  };

  get collectionAttributes() {
    return this.props.projectCollection.attributes;
  }

  render() {
    const projectCollection = this.props.projectCollection;
    if (!projectCollection) return null;

    return (
      <section key={projectCollection.id} className="bg-neutral05">
        <div className="container">
          <Header projectCollection={projectCollection} />
          <div className="project-collection-header__filter">
            <Filters
              filterChangeHandler={this.props.filterChangeHandler}
              initialState={this.props.initialState}
            />
          </div>
          <div className="entity-section-wrapper__details">
            <Utility.EntityCount
              pagination={this.props.pagination}
              singularUnit="project"
              pluralUnit="projects"
              countOnly
            />
          </div>
          <ProjectList.Grid
            authenticated={this.props.authentication.authenticated}
            favorites={get(this.props.authentication, "currentUser.favorites")}
            projects={this.props.projects}
            dispatch={this.props.dispatch}
            paginationClickHandler={this.props.paginationClickHandler}
            pagination={this.props.pagination}
          />
        </div>
      </section>
    );
  }
}
