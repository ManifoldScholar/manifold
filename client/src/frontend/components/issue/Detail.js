import React from "react";
import PropTypes from "prop-types";
import CollectionNavigation from "frontend/components/CollectionNavigation";
import EntityCollection from "frontend/components/entity/Collection/EntityCollection";
import EntityHero from "frontend/components/entity/Hero";
import ContentBlockList from "frontend/components/content-block-list/List";
import { Warning } from "frontend/components/content-block/parts";
import { useFromStore } from "hooks";
import { useFrontendModeContext } from "hooks";
import Authorize from "hoc/Authorize";

function Detail({ issue }) {
  const { isStandalone } = useFrontendModeContext();
  const settings = useFromStore("settings", "select");
  const libraryDisabled = settings?.attributes?.general?.libraryDisabled;

  return (
    <>
      <EntityHero.Issue entity={issue} />
      <Authorize entity={issue} ability="fullyRead" successBehavior="hide">
        <EntityCollection
          BodyComponent={() => <Warning.AccessDenied entity={issue} />}
        />
      </Authorize>
      <ContentBlockList entity={issue} />
      {!isStandalone && !libraryDisabled && (
        <CollectionNavigation entityType="journals" bgColor="neutral05" />
      )}
    </>
  );
}

Detail.displayName = "Issue.Detail";

Detail.propTypes = {
  issue: PropTypes.object.isRequired
};

export default Detail;
