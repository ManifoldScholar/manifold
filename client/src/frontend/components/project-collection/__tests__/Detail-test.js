import React from "react";
import renderer from "react-test-renderer";
import Detail from "../Detail";
import build from "test/fixtures/build";
import { wrapWithRouter, renderWithRouter } from "test/helpers/routing";

describe("Frontend.ProjectCollection.Detail component", () => {
  const projects = [build.entity.project("1"), build.entity.project("2")];
  const projectCollection = build.entity.projectCollection(
    "1",
    {},
    { projects }
  );

  it("renders correctly", () => {
    const component = renderer.create(
      wrapWithRouter(
        <Detail
          projectCollection={projectCollection}
          projects={projects}
          pagination={build.pagination()}
          paginationClickHandler={jest.fn}
          filterChangeHandler={jest.fn}
          authentication={{}}
        />
      )
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
