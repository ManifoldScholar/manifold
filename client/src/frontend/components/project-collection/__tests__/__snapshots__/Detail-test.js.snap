// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Frontend.ProjectCollection.Detail component renders correctly 1`] = `
<section
  className="bg-neutral05"
>
  <div
    className="entity-section-wrapper container"
  >
    <div
      className="entity-section-wrapper__heading section-heading"
    >
      <div
        className="main"
      >
        <i
          aria-hidden="true"
          className="manicon"
        >
          <svg
            aria-hidden={true}
            className="manicon-svg"
            height={56}
            viewBox="0 0 64 64"
            width={56}
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fillRule="evenodd"
            >
              <path
                d="M44.314549,33.1909886 L45.9318511,32.0144114 L49.1634511,36.4565114 L47.546149,37.6330886 L44.314549,33.1909886 Z M47.7134278,29.8115965 L48.5257723,27.9840035 L53.5454723,30.2152035 L52.7331278,32.0427965 L47.7134278,29.8115965 Z M48.9673311,25.1854584 L48.7570689,23.1965416 L54.2197689,22.6190416 L54.4300311,24.6079584 L48.9673311,25.1854584 Z M34.6515965,38.9963723 L32.8240035,38.1840278 L35.0552035,33.1643278 L36.8827965,33.9766722 L34.6515965,38.9963723 Z M42.2477586,39.6706708 L40.2588414,39.8809292 L39.6813414,34.4181292 L41.6702586,34.2078708 L42.2477586,39.6706708 Z M21.8055661,55.2600612 L20.0844339,56.2787388 L13.8494454,45.7442583 C10.1665975,39.5221317 11.8255838,31.5192434 17.6783357,27.2738322 L26.7793357,20.6722322 L27.9536643,22.2911678 L18.8526639,28.892768 C13.8357087,32.5319174 12.4136324,39.3919604 15.5705661,44.7255612 L21.8055661,55.2600612 Z M34.6171323,10.3538753 L26.4778605,16.2936667 L32.7831452,28.1653016 L43.999866,19.9791535 L34.6171323,10.3538753 Z M32.0988548,31.1406984 L23.8981395,15.7003333 L34.8388677,7.71612467 L47.042134,20.2348465 L32.0988548,31.1406984 Z M10.3684,56.7693 L10.3684,54.7693 L30.2014,54.7693 L30.2014,56.7693 L10.3684,56.7693 Z M30.9602576,11.6871433 L29.4425424,12.9896566 L26.6075424,9.6862566 L28.1252576,8.38374333 L30.9602576,11.6871433 Z"
                fill="currentColor"
                fillRule="nonzero"
              />
              <path
                d="M44.314549,33.1909886 L45.9318511,32.0144114 L49.1634511,36.4565114 L47.546149,37.6330886 L44.314549,33.1909886 Z M47.7134278,29.8115965 L48.5257723,27.9840035 L53.5454723,30.2152035 L52.7331278,32.0427965 L47.7134278,29.8115965 Z M48.9673311,25.1854584 L48.7570689,23.1965416 L54.2197689,22.6190416 L54.4300311,24.6079584 L48.9673311,25.1854584 Z M34.6515965,38.9963723 L32.8240035,38.1840278 L35.0552035,33.1643278 L36.8827965,33.9766722 L34.6515965,38.9963723 Z M42.2477586,39.6706708 L40.2588414,39.8809292 L39.6813414,34.4181292 L41.6702586,34.2078708 L42.2477586,39.6706708 Z M21.8055661,55.2600612 L20.0844339,56.2787388 L13.8494454,45.7442583 C10.1665975,39.5221317 11.8255838,31.5192434 17.6783357,27.2738322 L26.7793357,20.6722322 L27.9536643,22.2911678 L18.8526639,28.892768 C13.8357087,32.5319174 12.4136324,39.3919604 15.5705661,44.7255612 L21.8055661,55.2600612 Z M34.6171323,10.3538753 L26.4778605,16.2936667 L32.7831452,28.1653016 L43.999866,19.9791535 L34.6171323,10.3538753 Z M32.0988548,31.1406984 L23.8981395,15.7003333 L34.8388677,7.71612467 L47.042134,20.2348465 L32.0988548,31.1406984 Z M10.3684,56.7693 L10.3684,54.7693 L30.2014,54.7693 L30.2014,56.7693 L10.3684,56.7693 Z M30.9602576,11.6871433 L29.4425424,12.9896566 L26.6075424,9.6862566 L28.1252576,8.38374333 L30.9602576,11.6871433 Z"
                fill="currentColor"
                fillRule="nonzero"
              />
            </g>
          </svg>
        </i>
        <div
          className="body"
        >
          <h4
            className="title"
          >
            A Project Collection
          </h4>
        </div>
      </div>
    </div>
    <form
      className="form-list-filter entity-section-wrapper__tools"
      onSubmit={[Function]}
    >
      <div
        className="search-input"
      >
        <button
          className="search-button"
          type="submit"
        >
          <span
            className="screen-reader-text"
          >
            Search…
          </span>
          <svg
            aria-hidden={true}
            className="manicon-svg"
            height={20}
            viewBox="0 0 16 16"
            width={20}
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fillRule="evenodd"
            >
              <path
                d="M7.00000003,3 C4.79086102,3 3.00000001,4.79086101 3.00000001,7.00000002 C3.00000001,9.20913903 4.79086102,11 7.00000003,11 C9.20912238,11 10.999973,9.20916605 11.0000001,7.00005001 C10.9995314,4.79106172 9.20889466,3.00044165 7.00000003,3 Z M10.8715096,10.1643028 L13.9971535,13.2899467 L13.2900467,13.9970535 L10.1644164,10.8714232 C9.30247365,11.5768009 8.20065857,12.0000001 7.00000003,12.0000001 C4.23857626,12.0000001 2,9.76142379 2,7.00000002 C2,4.23857625 4.23857626,1.99999999 7.0001,2 C9.76123985,2.00055206 11.999448,4.23876021 12.0000001,6.99990005 C12.0000001,8.2005577 11.5768322,9.30236028 10.8715096,10.1643028 Z"
                fill="currentColor"
                fillRule="nonzero"
              />
              <path
                d="M7.00000003,3 C4.79086102,3 3.00000001,4.79086101 3.00000001,7.00000002 C3.00000001,9.20913903 4.79086102,11 7.00000003,11 C9.20912238,11 10.999973,9.20916605 11.0000001,7.00005001 C10.9995314,4.79106172 9.20889466,3.00044165 7.00000003,3 Z M10.8715096,10.1643028 L13.9971535,13.2899467 L13.2900467,13.9970535 L10.1644164,10.8714232 C9.30247365,11.5768009 8.20065857,12.0000001 7.00000003,12.0000001 C4.23857626,12.0000001 2,9.76142379 2,7.00000002 C2,4.23857625 4.23857626,1.99999999 7.0001,2 C9.76123985,2.00055206 11.999448,4.23876021 12.0000001,6.99990005 C12.0000001,8.2005577 11.5768322,9.30236028 10.8715096,10.1643028 Z"
                fill="currentColor"
                fillRule="nonzero"
              />
            </g>
          </svg>
        </button>
        <label
          className="screen-reader-text"
          htmlFor="filters-search-1"
        >
          Enter Search Criteria
        </label>
        <input
          id="filters-search-1"
          onChange={[Function]}
          placeholder="Search…"
          type="text"
          value=""
        />
      </div>
      <div
        className="select-group inline"
      >
        <div
          className="select"
        >
          <select
            onChange={[Function]}
            value=""
          >
            <option
              value=""
            >
              Sort
            </option>
            <option
              value="sort_title ASC"
            >
              A-Z
            </option>
            <option
              value="sort_title DESC"
            >
              Z-A
            </option>
          </select>
          <i
            aria-hidden="true"
            className="manicon manicon-caret-down"
          />
        </div>
      </div>
      <button
        className="reset-button"
        onClick={[Function]}
      >
        Reset Search + Filters
      </button>
    </form>
    <div
      className="entity-section-wrapper__details"
    >
      <p
        className="list-total"
      >
        <span>
          2
        </span>
         projects
      </p>
    </div>
    <nav
      className="project-list grid entity-section-wrapper__body"
    >
      <ul>
        <li>
          <a
            href="/projects/slug-1"
            onClick={[Function]}
          >
            <figure
              className="cover"
            />
            <div
              className="meta"
            >
              <h3
                className="name"
              >
                <span
                  className="title-text"
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "Rowan Test",
                    }
                  }
                />
              </h3>
            </div>
          </a>
        </li>
        <li>
          <a
            href="/projects/slug-2"
            onClick={[Function]}
          >
            <figure
              className="cover"
            />
            <div
              className="meta"
            >
              <h3
                className="name"
              >
                <span
                  className="title-text"
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "Rowan Test",
                    }
                  }
                />
              </h3>
            </div>
          </a>
        </li>
      </ul>
    </nav>
    <div
      className="entity-section-wrapper__pagination"
    >
      <nav
        className="list-pagination-primary"
      >
        <ul
          className={null}
        >
          <li
            className="pagination-previous disabled"
          >
            <a
              data-id="page-prev"
              href="#pagination-target"
              onClick={[MockFunction]}
            >
              <i
                aria-hidden="true"
                className="manicon manicon-arrow-long-left"
              />
              <span
                aria-hidden="true"
              >
                Prev
              </span>
              <span
                className="screen-reader-text"
              >
                Previous Page
              </span>
            </a>
          </li>
          <li
            className="active ordinal"
          >
            <a
              href="#pagination-target"
              onClick={[MockFunction]}
            >
              <span
                aria-hidden="true"
              >
                1
              </span>
              <span
                className="screen-reader-text"
              >
                Go to page: 
                1
              </span>
            </a>
          </li>
          <li
            className="ordinal"
          >
            <a
              href="#pagination-target"
              onClick={[MockFunction]}
            >
              <span
                aria-hidden="true"
              >
                2
              </span>
              <span
                className="screen-reader-text"
              >
                Go to page: 
                2
              </span>
            </a>
          </li>
          <li
            className="pagination-next "
          >
            <a
              data-id="page-next"
              href="#pagination-target"
              onClick={[MockFunction]}
            >
              <span
                className="screen-reader-text"
              >
                Next Page
              </span>
              <span
                aria-hidden="true"
              >
                Next
              </span>
              <i
                aria-hidden="true"
                className="manicon manicon-arrow-long-right"
              />
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</section>
`;
